src/audio/module_adapter/module_adapter.c

```diff
@@ -623,54 +630,126 @@ static void module_adapter_process_output(struct comp_dev *dev)
+		if (!sinks_c[i]->sink || sinks_c[i]->sink->state != dev->state) {
+			i++;
+			continue;
+		}
+
+		comp_get_copy_limits_frame_aligned(source_c[0], sinks_c[i], &c);
+
+		min_frames = MIN(min_frames, c.frames);
+		source_frame_bytes = c.source_frame_bytes;
+
+		mod->output_buffers[num_output_buffers].size = 0;
+		mod->output_buffers[num_output_buffers].data = &sinks_c[i]->stream;
+		num_output_buffers++;
+		i++;
+	}
+
+	buffer_stream_invalidate(source_c[0], min_frames * source_frame_bytes);
+	/* note that the size is in number of frames not the number of bytes */
+	mod->input_buffers[0].size = min_frames;
+	mod->input_buffers[0].consumed = 0;
+	mod->input_buffers[0].data = &source_c[0]->stream;
+
+	return num_output_buffers;
+}
+
 int module_adapter_copy(struct comp_dev *dev)
 {
 	struct processing_module *mod = comp_get_drvdata(dev);
 	struct module_data *md = &mod->priv;
 	struct comp_buffer *source, *sink;
 	struct comp_buffer __sparse_cache *source_c[PLATFORM_MAX_STREAMS];
+	struct comp_buffer __sparse_cache *sinks_c[PLATFORM_MAX_STREAMS];
 	struct comp_buffer __sparse_cache *sink_c = NULL;
-	struct comp_copy_limits c;
 	struct list_item *blist;
 	size_t size = MAX(mod->deep_buff_bytes, mod->period_bytes);
 	uint32_t min_free_frames = UINT_MAX;
 	uint32_t num_input_buffers = 0;
+	uint32_t num_output_buffers = 0;
 	int ret, i = 0;
 
 	comp_dbg(dev, "module_adapter_copy(): start");
 
 	/*
 	 * Simplify calculation of bytes_to_process for modules that produce period_bytes every
-	 * period and have N sources and only 1 sink buffer
+	 * period and have a 1:1, 1:N or N:1 source:sink buffer configuration
 	 */
 	if (mod->simple_copy) {
-		sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
-		sink_c = buffer_acquire(sink);
+		list_for_item(blist, &dev->bsink_list) {
+			sink = container_of(blist, struct comp_buffer, source_list);
+			sinks_c[i++] = buffer_acquire(sink);
+		}
+		i = 0;
 		list_for_item(blist, &dev->bsource_list) {
 			source = container_of(blist, struct comp_buffer, sink_list);
-			source_c[i] = buffer_acquire(source);
-
-			/* check if the source dev is in the same state as the dev */
-			if (!source_c[i]->source || source_c[i]->source->state != dev->state) {
-				i++;
-				continue;
-			}
-
-			comp_get_copy_limits_frame_aligned(source_c[i], sink_c, &c);
-
-			buffer_stream_invalidate(source_c[i], c.frames * c.source_frame_bytes);
-
-			/* note that the size is in number of frames not the number of bytes */
-			mod->input_buffers[num_input_buffers].size = c.frames;
-			mod->input_buffers[num_input_buffers].consumed = 0;
-
-			mod->input_buffers[num_input_buffers].data = &source_c[i]->stream;
-			num_input_buffers++;
-			i++;
+			source_c[i++] = buffer_acquire(source);
+		}
+		if (mod->num_output_buffers == 1) {
+			num_input_buffers = module_single_sink_setup(dev, source_c, sinks_c);
+			if (sinks_c[0]->sink && sinks_c[0]->sink->state == dev->state)
+				num_output_buffers = 1;
```

Comment:
so what does it mean in this case `if (!sinks_c[0]->sink)`?

Reply:
It means the sink device is not connected or it is not active yet. This check is quite important for the mixin

Reply:
the device isn't connected but the buffer is already there?

Reply:
The first condition is unlikely @lyakh . It's only there as a paranoia check.

Reply:
Sorry, I'm not a big fan of such "paranoia checks." I understand this software is complex and probably no single person in the world can know all the paths that can get executed by it. In cases like this maybe at least we can put an explicit check if we don't know if that can happen or not: if we don't want to make this an assert and risk an exception, at least just adding an error print could be enough.